File: CSharpRewrite/CSharpRewrite/Assignments/Extra/Road.cs
namespace CSharpRewrite.Assignments.Extra;

public static class Road
{
    private readonly struct Route : IComparable<Route>
    {
        public int Destination { get; init; }
        public IEnumerable<(char, int)> Path { get; init; }

        public int CompareTo(Route other)
        {
            var destinationCompareResult = Destination.CompareTo(other.Destination);
            if (destinationCompareResult != 0)
                return destinationCompareResult;

            var zipped = Path.Zip(other.Path);
            foreach (var (first, second) in zipped)
            {
                var compareResult = first.CompareTo(second);
                if (compareResult != 0)
                    return compareResult;
            }

            return Path.Count().CompareTo(other.Path.Count());
        }
    }

    private struct RoutingResult
    {
        public Route A { get; init; }
        public Route B { get; init; }
    }
    
    public static IEnumerable<(char, int)> Run()
    {
        var content = File.ReadAllLines("road.txt");
        var map = ParseMap(content);
        var values = GroupValues(new Queue<int>(map));
        return OptimalPath(values);
    }

    private static IEnumerable<int> ParseMap(IEnumerable<string> content)
        => content.Select(s => s.Trim()).Select(int.Parse);

    private static Queue<(int, int, int)> GroupValues(Queue<int> numbers)
    {
        var queue = new Queue<(int, int, int)>();
        while (numbers.Count != 0)
        {
            var a = numbers.Dequeue();
            var b = numbers.Dequeue();
            var x = numbers.Dequeue();
            queue.Enqueue((a, b, x));
        }

        return queue;
    }

    private static RoutingResult ShortestSteps(RoutingResult routingResult, (int, int, int) item)
    {
        var (a, b, x) = item;
        var optA1 = new Route()
        {
            Destination = routingResult.A.Destination + a,
            Path = routingResult.A.Path.Append(('a', a))
        };

        var optA2 = new Route()
        {
            Destination = routingResult.B.Destination + b + x,
            Path = routingResult.B.Path.Append(('b', b)).Append(('x', x))
        };
        
        var optB1 = new Route()
        {
            Destination = routingResult.B.Destination + b,
            Path = routingResult.B.Path.Append(('b', b))
        };
        
        var optB2 = new Route()
        {
            Destination = routingResult.A.Destination + a + x,
            Path = routingResult.A.Path.Append(('a', a)).Append(('x', x))
        };

        return new RoutingResult()
        {
            A = new[] { optA1, optA2 }.Min(),
            B = new[] { optB1, optB2 }.Min()
        };
    }

    private static IEnumerable<(char, int)> OptimalPath(IEnumerable<(int, int, int)> values)
    {
        var result = values.Aggregate(new RoutingResult()
        {
            A = new Route() { Destination = 0, Path = Enumerable.Empty<(char, int)>() },
            B = new Route() { Destination = 0, Path = Enumerable.Empty<(char, int)>() },
        }, ShortestSteps);

        if (result.A.Path.Last() != ('x', 0))
            return result.A.Path;
        return result.B.Path.Last() != ('x', 0) ? result.B.Path : new Route().Path;
    }
}

File: ScalaRewrite/src/main/scala/assignments/extra/Road.scala
package assignments.extra

import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Path
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer

case class PathItem(name: Char, cost: Int) extends Comparable[PathItem]:
  override def compareTo(o: PathItem): Int =
    cost.compareTo(o.cost)

case class Route(destination: Int, path: mutable.ArrayBuffer[PathItem]) extends Comparable[Route]:
  override def compareTo(o: Route): Int =
    val destinationCompareResult = destination.compareTo(o.destination)
    if destinationCompareResult != 0 then
      destinationCompareResult
    else
      val zipped = path.zip(o.path)
      for
        item <- zipped
      do
        val result = item._1.compareTo(item._2)
        if result != 0 then
          return result
      path.lengthCompare(o.path)

case class RoutingResult(a: Route, b: Route)

def runRoadAnalysis(): mutable.ArrayBuffer[PathItem] =
  val content = Files.readString(Path.of("assets/road.txt"), Charset.forName("utf-8"))
  val parsedMap = parseMap(content)
  val groupedValues = groupValues(parsedMap)
  optimalPath(groupedValues)

private def parseMap(content: String): mutable.ArrayDeque[Int] =
  val split = content
    .split(Array(' ', '\n', '\t', '\r'))
    .map(s => s.trim)
    .filter(s => !s.isEmpty && s.length > 0)
    .map(s => s.toInt)
  mutable.ArrayDeque.from(split)

private def groupValues(numbers: mutable.ArrayDeque[Int]): mutable.ArrayDeque[(Int, Int, Int)] =
  val result = mutable.ArrayDeque.empty[(Int, Int, Int)]
  val numbersToMove = numbers
  while (!numbersToMove.isEmpty) do
    val a = numbersToMove.removeHeadOption().getOrElse(0)
    val b = numbersToMove.removeHeadOption().getOrElse(0)
    val x = numbersToMove.removeHeadOption().getOrElse(0)
    result += ((a, b, x))
  result

private def shortestSteps(acc: RoutingResult, elem: (Int, Int, Int)): RoutingResult =
  val ((a, b, x)) = elem
  val pathA1 = acc.a.path.clone()
  pathA1 += PathItem('a', a)
  val optA1 = Route(acc.a.destination + a, pathA1)

  val pathA2 = acc.b.path.clone()
  pathA2 += PathItem('b', b)
  pathA2 += PathItem('x', x)
  val optA2 = Route(acc.b.destination + b + x, pathA2)

  val pathB1 = acc.b.path.clone()
  pathB1 += PathItem('b', b)
  val optB1 = Route(acc.b.destination + b, pathB1)

  val pathB2 = acc.a.path.clone()
  pathB2 += PathItem('a', a)
  pathB2 += PathItem('x', x)
  val optB2 = Route(acc.a.destination + a + x, pathB2)

  RoutingResult(if optA1.compareTo(optA2) < 0 then optA1 else optA2,
    if optB1.compareTo(optB2) < 0 then optB1 else optB2)

private def optimalPath(values: mutable.ArrayDeque[(Int, Int, Int)]): mutable.ArrayBuffer[PathItem] =
  val initialResult = RoutingResult(Route(0, ArrayBuffer.empty[PathItem]), Route(0, ArrayBuffer.empty[PathItem]))
  val result = values.foldLeft(initialResult)(shortestSteps)
  if result.a.path.last != PathItem('x', 0) then
    result.a.path
  else if result.b.path.last != PathItem('x', 0) then
    result.b.path
  else
    ArrayBuffer.empty